### DISCUSSION TITLE
Slow opening and clipping of files

### DISCUSSION THREAD

**Discussion by nkorinek** [corteva/rioxarray#234] [Q&A] [answered]
Hey again!

Sorry for the back to back questions, hope it's not to much trouble! I've just been updating a lot of our old code to use rioxarray instead of rasterio which has lead to a few questions.

We have a notebook that loops through 46 Landsat images, opens them, clips them, and masks out cloud values. When we use rasterio to do this, the whole notebook can run in under one second. With rioxarray, it takes over 30 seconds. I timed them. It seems that the slow part is opening and clipping all those files in rioxarray. By commenting out line by line, these lines seem to be the ones that slows down the code the most

```
band_4 = rxr.open_rasterio(raster_list[3], masked=True).squeeze()
band_4_crop_data = band_4.rio.clip(site_bound.geometry.apply(mapping))
```

I realize these are somewhat complex functions, but the same thing was done in rasterio very quickly. Do you have any hint as to why this may be taking so long? Any help would be appreciated! If you need more details I can help provide those too

**Comment by snowman2** [comment, marked as answer]
There are several ways to speed things up.

1. Use `from_disk=True` when clipping. This will reduce the amount of data loaded in. (rioxarray 0.2+)
2. With rioxarray 0.2+ and rasterio 1.2+ you don't need `.apply(mapping)` anymore.
3. Enable global context with pyproj 3+  with rioxarray < 0.3 (https://pyproj4.github.io/pyproj/stable/api/global_context.html#global-context) - be very careful with this option. With rioxarray 0.3+ use `rioxarray.set_options(export_grid_mapping=False)`

Putting it all together:
```python
import os
from glob import glob

import earthpy.spatial as es
import numpy as np
import pandas as pd
import geopandas as gpd
import rioxarray as rxr
import pyproj

# use for rioxarray < 0.3
# pyproj.set_use_global_context(True)

# Get a list of each directory
path = os.path.join("ndvi-automation", "sites")

# Get a list of each directory
sites = glob(path + "/*/")

# Define columns and list used to create pandas dataframe
col_names = ["site", "date", "mean_ndvi"]
output = []

for site in sites:

    # Get site name
    site_name = os.path.basename(os.path.normpath(site))

    # Define directories
    landsat_dir = os.path.join(site, "landsat-crop")
    vector_dir = os.path.join(site, "vector")

    # Define site boundary
    site_boundary_path = os.path.join(
        vector_dir,  site_name + "-crop.shp")
    site_bound = gpd.read_file(site_boundary_path)

    # Get list of landsat scene directories
    cropped_directories = sorted(glob(os.path.join(landsat_dir, "LC08*")))

    # Calculate NDVI for each directory
    for directory in cropped_directories:

        # Get date from directory name
        dir_head, dir_tail = os.path.split(directory)
        date = dir_tail[10:18]

       # Open bands
        raster_list = sorted(glob(os.path.join(directory, "*band*.tif")))
        with rioxarray.set_options(export_grid_mapping=False):
            band_4_crop_data = rxr.open_rasterio(
                raster_list[3], masked=True,
            ).rio.clip(
                site_bound.geometry, from_disk=True,
            ).squeeze()
    
            band_5_crop_data = rxr.open_rasterio(
                raster_list[4], masked=True,
            ).rio.clip(
                site_bound.geometry, from_disk=True,
            ).isel(band=0)

        band_4_crop_data = band_4_crop_data.where(
            ~((band_4_crop_data < 0) | (band_4_crop_data > 10000))
        )
        band_5_crop_data = band_5_crop_data.where(
            ~((band_5_crop_data < 0) | (band_5_crop_data > 10000))
        )

        # Calculate NDVI for site using masked data
        ndvi_landsat = es.normalized_diff(
            band_5_crop_data.values.astype('f4'), band_4_crop_data.values.astype('f4'),
        )

        # Add date and NDVI mean to list
        output.append([site_name, date, np.nanmean(ndvi_landsat)])

    # Convert list to pandas dataframe and save to csv
    ndvi_ts_unclean = pd.DataFrame(output, columns=col_names)
    out_path_no_clean = os.path.join(
        "outputs", "ndvi_2017_not_clean.csv")
    ndvi_ts_unclean.to_csv(out_path_no_clean)
    ndvi_df_unclean = pd.read_csv(out_path_no_clean,
                          parse_dates=['date'],
                          na_values=[-9999],
                          index_col=['date'])
```

**Reply by snowman2** [reply]
@nkorinek have you had a chance to try it out? I am curious how the performance compares for you with these adjustments.

**Reply by lwasser** [reply]
@snowman2 thank you for this. i've been playing with it and will do some speed tests. I haven't had a chance yet to use it given i think that i need to update our environments. do we need rioxarray 0.2 or greater? i think that may be my problem. if that is the case i may update all of our environments as well

**Reply by lwasser** [reply]
also that global context item - i'm reading "proceed with caution" in the docs. what is that doing and what are the "risks" associated with using it? thank you for all of the help below. we are playing and doing lots of speed tests.

**Reply by snowman2** [reply]
> do we need rioxarray 0.2 or greater

I would recommend rioxarray 0.2+ and rasterio 1.2+. Otherwise steps 1 & 2 won't work.

> what is that doing and what are the "risks" associated with using it?

It is not thread safe. If you are using this application using multiple threads (i.e. dask) it will break.

**Reply by snowman2** [reply]
With changes in master (#241) and using `rioxarray.set_options(export_grid_mapping=False)` as shown in the updated example, I get speed similar to `pyproj.set_use_global_context(True)`

**Reply by nkorinek** [reply]
Oooh, good to know!

**Reply by nkorinek** [reply]
I tested this in our speed testing notebook and got significant speed increases! Seems to be about twice as fast, based on my timing, which is great! Thanks for putting this in here! Quick question: does this run the same risks as `pyproj.set_use_global_context(True)` or is this safer to run?

**Reply by snowman2** [reply]
> is this safer to run?

This is much safer to run.

**Reply by lwasser** [reply]
this i s  fantastic!  @nkorinek thank you  for  following  up  here.  @snowman2 when  will this new fix `rioxarray.set_options(export_grid_mapping=False)` be  released? We  will then   update lessons and environments   accordingly!

**Reply by snowman2** [reply]
Current plan is pretty soon. I will post here when it is released so you can update.

**Reply by lwasser** [reply]
wonderful. no rush  on our   end  but once  we know  we can update.  Thank you  @snowman2 !!  we've learned  a  lot and i  love being able to  teach  approaches that  are  efficient  so we appreciate  your time in working through this with us.

**Reply by snowman2** [reply]
0.3.0 is on pypi :rocket:

**Reply by lwasser** [reply]
yes!!  thank you. We will do some updates on our end!

**Comment by nkorinek** [comment]
Hey! Just tried out the changes, sadly they didn't make too big of a difference. Before the changes I ran the notebook and it took 27 seconds to run. After the changes it took 25. So faster! But not by a noticeable amount.

**Reply by snowman2** [reply]
That's unfortunate. If you could provide:
1. The input files
2. The rioxarray code snipped
3. The raw rasterio code snippet
That would be helpful to dig into it later.

**Reply by nkorinek** [reply]
Ya! Here is the data we are using: https://ndownloader.figshare.com/files/13431344 . The students loop through the directories and opens/clips all of the rasters. Here is a snippet of the relevant code in rioxarray: 
```

# Get a list of each directory
path = os.path.join("ndvi-automation", "sites")

# Get a list of each directory
sites = glob(path + "/*/")

# Define columns and list used to create pandas dataframe
col_names = ["site", "date", "mean_ndvi"]
output = []

for site in sites:

    # Get site name
    site_name = os.path.basename(os.path.normpath(site))

    # Define directories
    landsat_dir = os.path.join(site, "landsat-crop")
    vector_dir = os.path.join(site, "vector")

    # Define site boundary
    site_boundary_path = os.path.join(
        vector_dir,  site_name + "-crop.shp")
    site_bound = gpd.read_file(site_boundary_path)

    # Get list of landsat scene directories
    cropped_directories = glob(os.path.join(landsat_dir, "LC08*"))
    cropped_directories.sort()

    # Calculate NDVI for each directory
    for directory in cropped_directories:

        # Get date from directory name
        dir_head, dir_tail = os.path.split(directory)
        date = dir_tail[10:18]

       # Open bands
        raster_list = glob(os.path.join(directory, "*band*.tif"))
        raster_list.sort()

        band_4 = rxr.open_rasterio(raster_list[3], masked=True, parse_coordinates=False).squeeze()
        band_4_crop_data = band_4.rio.clip(site_bound.geometry, from_disk=True)

        band_4_mask = ((band_4_crop_data < 0) | (band_4_crop_data > 10000))
        band_4_crop_data = band_4_crop_data.where(~xr.where(band_4_mask, True, False))
        
        band_5 = rxr.open_rasterio(raster_list[4], masked=True, parse_coordinates=False).squeeze()
        band_5_crop_data = band_5.rio.clip(site_bound.geometry, from_disk=True)
        
        band_5_mask = ((band_5_crop_data < 0) | (band_5_crop_data > 10000))
        band_5_crop_data = band_5_crop_data.where(~xr.where(band_5_mask, True, False))
        
        # Calculate NDVI for site using masked data
        ndvi_landsat = es.normalized_diff(
            band_5_crop_data.values.astype('f4'), band_4_crop_data.values.astype('f4'))

        # Add date and NDVI mean to list
        output.append([site_name, date, np.nanmean(ndvi_landsat)])

# Convert list to pandas dataframe and save to csv
ndvi_ts_unclean = pd.DataFrame(output, columns=col_names)
out_path_no_clean = os.path.join(
    "ndvi-automation", "outputs", "ndvi_2017_not_clean.csv")
ndvi_ts_unclean.to_csv(out_path_no_clean)
ndvi_df_unclean = pd.read_csv(out_path_no_clean,
                      parse_dates=['date'],
                      na_values=[-9999],
                      index_col=['date'])

ndvi_df_unclean
```
And here is a snippet of the same code in rasterio:
```
# Get a list of each directory
path = os.path.join("data", "ndvi-automation", "sites")

# Get a list of each directory
sites = glob(path + "/*/")

# Define columns and list used to create pandas dataframe
col_names = ["site", "date", "mean_ndvi"]
output = []

for site in sites:

    # Get site name
    site_name = os.path.basename(os.path.normpath(site))

    # Define directories
    landsat_dir = os.path.join(site, "landsat-crop")
    vector_dir = os.path.join(site, "vector")

    # Define site boundary
    site_boundary_path = os.path.join(
        vector_dir,  site_name + "-crop.shp")
    site_bound = gpd.read_file(site_boundary_path)

    # Get list of landsat scene directories
    cropped_directories = glob(os.path.join(landsat_dir, "LC08*"))
    cropped_directories.sort()

    # Calculate NDVI for each directory
    for directory in cropped_directories:

        # Get date from directory name
        dir_head, dir_tail = os.path.split(directory)
        date = dir_tail[10:18]

       # Open bands
        raster_list = glob(os.path.join(directory, "*band*.tif"))
        raster_list.sort()

        with rio.open(raster_list[3]) as band_4:
            band_4_crop, band_4_meta = es.crop_image(
                band_4, site_bound)
            mask = (band_4_crop < 0) | (band_4_crop > 10000)
            band_4_crop_data = ma.masked_array(band_4_crop, mask)

        with rio.open(raster_list[4]) as band_5:
            band_5_crop, band_5_meta = es.crop_image(
                band_5, site_bound)
            mask = (band_5_crop < 0) | (band_5_crop > 10000)
            band_5_crop_data = ma.masked_array(band_5_crop, mask)

        # Calculate NDVI for site using masked data
        ndvi_landsat = es.normalized_diff(
            band_5_crop_data.astype('f4'), band_4_crop_data.astype('f4'))

        # Add date and NDVI mean to list
        output.append([site_name, date, np.nanmean(ndvi_landsat)])

# Convert list to pandas dataframe and save to csv
ndvi_ts_unclean = pd.DataFrame(output, columns=col_names)
out_path_no_clean = os.path.join(
    "data", "ndvi-automation", "outputs", "ndvi_2017_not_clean.csv")
ndvi_ts_unclean.to_csv(out_path_no_clean)
```
This is done 3 times through the notebook (since these are homework notebooks). If you need the full notebook to make it easier to trouble shoot I can send those over to you too! Thanks!

**Reply by snowman2** [reply]
Thanks for sharing - that will help a bit. Can you try doing `squeeze` after the `clip` operation? Also, formatted code is much better than screenshots, Please swap it out if you wouldn't mind.

**Reply by nkorinek** [reply]
Definitely, apologies! I updated the question to use formatted code instead of screenshots. Also: I tried with squeeze after the clip operation. This time it ran in 23 seconds instead of 25! Getting more efficient bit by bit!

**Reply by snowman2** [reply]
The culprit was `pyproj`. When I enabled the global context (https://pyproj4.github.io/pyproj/stable/api/global_context.html#global-context), I was able to get it down to ~2 seconds. But, you can only enable the global context for single threaded applications. I updated my answer above.

**Reply by nkorinek** [reply]
This one did reduce the time a lot (8 seconds for the whole notebook), thanks so much for your help these last couple of days Alan, I really appreciate it!

**Reply by snowman2** [reply]
Glad to hear it worked for you :+1: Side note: Did you notice the minor change in the masking logic I made as well?

**Reply by nkorinek** [reply]
Ahh, I missed that! Makes it much cleaner, thanks!

**Reply by lwasser** [reply]
what version of pyproj do we need to ensure this all runs properly as well?

**Comment by lwasser** [comment]
ok let me try to summarize the changes that i see here

1. i think we need rioxarray 0.2.0 or greater for this and pyproj version ???
2. enabling `pyproj.set_use_global_context(True)` speeds things up but there are some dangers here that i'd  like to better understand before using it and teaching it. 
3. When you are using clip you are calling `from_disk` - this i think requires a new  version of rioxarray. what does this do to enhance speed for us? i'll test it after updating our envts today!

```
 band_4_crop_data = rxr.open_rasterio(
            raster_list[3], masked=True,
        ).rio.clip(
            site_bound.geometry, from_disk=True,
        ).squeeze()
```

# i also see this approach
```
       band_5_crop_data = rxr.open_rasterio(
            raster_list[4], masked=True,
        ).rio.clip(
            site_bound.geometry, from_disk=True,
        ).isel(band=0)
```
is it better to use squeeze or select  a band? squeeze seems to work nicely! 

In short does my comment above summarize the  tweeks to make things faster?

* from_disk and
* pyproj 
being the two big changes to enhance speed but both depend upon newer versions of tools.

**Reply by snowman2** [reply]
From: https://github.com/corteva/rioxarray/discussions/234#discussioncomment-343391
```
1. Use from_disk=True when clipping. This will reduce the amount of data loaded in. (rioxarray 0.2+)
2. With rioxarray 0.2+ and rasterio 1.2+ you don't need .apply(mapping) anymore.
3. Enable global context with pyproj 3+ (https://pyproj4.github.io/pyproj/stable/api/global_context.html#global-context) - be very careful with this option.

```

**Reply by snowman2** [reply]
> is it better to use squeeze or select a band? squeeze seems to work nicely!

Either one is fine.

**Reply by lwasser** [reply]
ok great  - i just did a local envt update. the pyproj 3x  doesn't seem to be installing by default - i will  see if i can  force it without  breaking our environment (the joys  of python environments :)) ! much appreciated @snowman2 i'll share some cleaned up  code with times just for a record here when we are done! (it is very close) Your help is  much appreciated!

**Reply by lwasser** [reply]
If this is helpful - for kicks i ran a few speed tests on the open and clip series isolated:

<img width="750" alt="Screen Shot 2021-02-10 at 11 33 00 AM" src="https://user-images.githubusercontent.com/7649194/107554887-c25a6280-6b93-11eb-8e41-5d24c3aeb8ff.png">

i do see some small improvements using both `from_disk` and the `pyproj` fix.

**Reply by snowman2** [reply]
@lwasser, for your tests, I recommend two changes:

1. Use `pyproj.set_use_global_context(False)` for the other examples to be sure the setting is off before running.
2. I recommend doing the clip immediately after opening to ensure you have the file handle (so do the squeeze after clip) [ref](https://corteva.github.io/rioxarray/stable/examples/clip_geom.html#Clipping-larger-rasters).

**Comment by lwasser** [comment]
[I  created a summary notebook that shows (significant!!) speed gains using from_disk  and pyproj global context (at the bottom of the notebook there  are timed sections). ](https://gist.github.com/lwasser/4e63e9efd25f3aa558b00fd781a66de2). i do plan to  add this as a lesson to our website as  well!! thank you for the help.

i  read through the pyproj documentation but still don't  understand why that  global  context option is "dangerous".  Is there something else that i can read  to  better understand  what the dangers are? Does it somehow use more processing power (threads?). i may just not understand what it's doing.

**Reply by snowman2** [reply]
> still don't understand why that global context option is "dangerous"

Essentially you could get unexpected behavior or crash your program if you use multi-threading with the global context enabled.
